crates/containerd-shim-wasm/src/sandbox/shim.rs

```diff
@@ -1,1274 +0,0 @@
-        s.kill()?;
-
-        self.instance.kill(signal)
-    }
-
-    fn delete(&self) -> Result<()> {
-        let mut s = self.state.write().unwrap();
-        s.delete()?;
-
-        let res = self.instance.delete();
-
-        if res.is_err() {
-            // Always `Ok(())` because we hold the lock since `s.delete()`
-            let _ = s.stop();
-        }
-
-        res
-    }
-
-    fn wait(&self) -> (u32, DateTime<Utc>) {
-        let res = self.instance.wait();
-        let mut s = self.state.write().unwrap();
-        *s = TaskState::Exited;
-        res
-    }
-
-    fn wait_timeout(&self, t: impl Into<Option<Duration>>) -> Option<(u32, DateTime<Utc>)> {
-        let res = self.instance.wait_timeout(t);
-        if res.is_some() {
-            let mut s = self.state.write().unwrap();
-            *s = TaskState::Exited;
-        }
-        res
-    }
-}
-
-type EventSender = Sender<(String, Box<dyn MessageDyn>)>;
-
-#[derive(Debug, Clone, Copy)]
-enum TaskState {
-    Created,
-    Starting,
-    Started,
-    Exited,
-    Deleting,
-}
-
-impl TaskState {
-    pub fn start(&mut self) -> Result<()> {
-        *self = match self {
-            Self::Created => Ok(Self::Starting),
-            _ => state_transition_error(*self, Self::Starting),
-        }?;
-        Ok(())
-    }
-
-    pub fn kill(&mut self) -> Result<()> {
-        *self = match self {
-            Self::Started => Ok(Self::Started),
-            _ => state_transition_error(*self, "Killing"),
-        }?;
-        Ok(())
-    }
-
-    pub fn delete(&mut self) -> Result<()> {
-        *self = match self {
-            Self::Created | Self::Exited => Ok(Self::Deleting),
-            _ => state_transition_error(*self, Self::Deleting),
-        }?;
-        Ok(())
-    }
-
-    pub fn started(&mut self) -> Result<()> {
-        *self = match self {
-            Self::Starting => Ok(Self::Started),
-            _ => state_transition_error(*self, Self::Started),
-        }?;
-        Ok(())
-    }
-
-    pub fn stop(&mut self) -> Result<()> {
-        *self = match self {
-            Self::Started | Self::Starting => Ok(Self::Exited),
-            // This is for potential failure cases where we want delete to be able to be retried.
-            Self::Deleting => Ok(Self::Exited),
-            _ => state_transition_error(*self, Self::Exited),
-        }?;
-        Ok(())
-    }
-}
-
-fn state_transition_error<T>(from: impl std::fmt::Debug, to: impl std::fmt::Debug) -> Result<T> {
-    Err(Error::FailedPrecondition(format!(
-        "invalid state transition: {from:?} => {to:?}"
-    )))
-}
-
-type LocalInstances<T> = Arc<RwLock<HashMap<String, Arc<InstanceData<T>>>>>;
-
-/// Local implements the Task service for a containerd shim.
-/// It defers all task operations to the `Instance` implementation.
-#[derive(Clone)]
-pub struct Local<T: Instance + Send + Sync> {
-    engine: T::Engine,
-    instances: LocalInstances<T>,
-    events: Arc<Mutex<EventSender>>,
```

Comment:
The `Mutex` type was removed in the `RemoteEventSender` implementation, why?

Reply:
It was never needed to start with.
The new `EventSender` trait requires `Sync`, which means it's safe to call it's methods from different threads.
The previous `EventSender` type was also `Sync`, so it didn't need the `Mutex` either.